Bluepill-cli/Bluepill-cli/Simulator/SimulatorMonitor.m

```diff
@@ -199,22 +202,30 @@ - (void)onOutputReceived:(NSString *)output {
     __weak typeof(self) __self = self;
 
     // App crashed
-    if ([output isEqualToString:@"BP_APP_PROC_ENDED"] && __self.simulatorState == TestsStarted) {
-        [BPUtils printInfo:CRASH withString:@"%@/%@ crashed app.",
-                                             (self.currentClassName ?: self.previousClassName),
-                                             (self.currentTestName ?: self.previousTestName)];
-        [self stopTestsWithErrorMessage:@"App Crashed"
-                              forTestName:(self.currentTestName ?: self.previousTestName)
-                                  inClass:(self.currentClassName ?: self.previousClassName)];
-        self.exitStatus = BPExitStatusAppCrashed;
-        [[BPStats sharedStats] addApplicationCrash];
+    if ([output isEqualToString:@"BP_APP_PROC_ENDED"]) {
+        __self.parserState = Completed;
+        if (__self.testsState == Running || __self.testsState == Idle) {
+            if (__self.testsState == Running) {
```

Comment:
add tests to make sure we don't screw things up later

Reply:
There are tests already.

Reply:
I don't see any test for `self.appState`, `self.parserState` and `self.testsState`

Reply:
The tests that were already there test these boundary conditions. They were racy before, they should _always_ pass now.

Reply:
My point is with the introduction of so many state variables, how can you prevent making mistakes in future commits?

Reply:
I can't, and we probably need more testing... however the white-box testing you suggest isn't easy here so I'm relying in the black-box testing that was already there. If you can suggest a better way, I'm all ears.

Reply:
What if just check the parserState and testsState in testParsingCrash test, to make sure they are both 'completed', because that is when you set these two states?

Reply:
Yeah, I looked at something like that and it's not ideal. That's a bit just testing that the code does what the code says... the main problem for white-box testing is that the code is not structured in a way that you can mock things out and pretend you have a Simulator or an App running and simulate failure. Our options are to rewrite a lot of code to make it more testable or rely on black-box testing which is what I'm doing. 

Since this is fixing a real problem where we misreport failures as successes I don't want to do a huge chunk of work or delay the fix.

Reply:
Agreed. Looks good to me unless @oliverhu has better ideas for writing the test.

Reply:
Not sure why it is hard to white-box test this with mocks ? It is super easy with OCMock. https://github.com/erikdoe/ocmock.

As a sense of urgency, we can always lean on the branch before switching back to master branch for internal usages. This is not fixing a real problem for now, but fixing a problem for the rest of time.

Reply:
Ah, I think the problem is you don't understand the urgency. `master` is busted right now, it'll report all is fine when in fact some tests have failed. It'll also report the app crashed when in fact it just hit the race this fixes. 

The reporting is busted both in `master` and in this PR, I'm working on fixing that but I figured it was high priority to, you know, at least exit with the right exit status.

Reply:
> Not sure why it is hard to white-box test this with mocks ? It is super easy with OCMock. https://github.com/erikdoe/ocmock.

Cool, I'll open an issue and assign it to you. Thank you for your contributions!

Reply:
It is common sense to fix the problem and also fix the root cause (why this could be broken) in the same PR. It is not proper to say, we fix this shit, don't care about the root cause, and break it again later, again and again, in the same loop. 

I understand your urgency, but please be a good citizen. Thank you for contribution as well!

Reply:
I don't think you get it so let me try again. `master` was okay as of around c330f51. Then I pushed some bad juju when I was trying to fix the problems with the missing dynamic libraries, JUnit report, etc. Now it's broken. You can dig out and revert the commits, push this PR, or do nothing and leave it broken. I'm not going to keep pushing on this PR and will just use the branch and keep fixing stuff but IMHO we owe it to the users of Bluepill to not have a broken master.

Reply:
We are on the same page for getting this commit in as soon as possible to fix the broken master (not released version). IMHO, it takes less than 30 minutes to add tests to cover this scenario with some white-box tests and since you have most context on this bug, it would even easier for you to add that. If you think the cost to you to add that is _far_ beyond that, I'm *OK* to merge this PR first, and follow up on #237 and add examples there.

Reply:
That sounds good to me. We definitively need more tests and easier white-box testing. As for the tests, the tests we discussed adding wouldn’t really test much more than “the code is doing what the code is doing”. Writing useful tests is much more than a 30 min job.